rangify readLink and symlink

aG0aep6G opened the pull request:
Removing @safe from the signatures so that attribute inference takes
care of things. There are @safe unittests already, ensuring that the
functions themselves don't do anything unsafe.

Replacing @trusted declarations with @trusted function literals. This
way it's more clear that the specific call is @trusted, not the called
function.

JakobOvrum commented on std/file.d:
```diff
@@ -2219,15 +2219,31 @@ unittest
+        auto oz = original.tempCString();
+        auto lz = link.tempCString();
+        alias posixSymlink = core.sys.posix.unistd.symlink;
+        immutable int result = () @trusted {return posixSymlink(oz, lz);} ();
```
Nitpick, please put space both before and after `{` and `}`

JakobOvrum commented on std/file.d:
```diff
@@ -2285,23 +2307,22 @@ version(Posix) @safe unittest
+    if (isInputRange!R && isSomeChar!(ElementEncodingType!R) &&
+        !isConvertibleToString!R);
+else version(Posix) string readLink(R)(R link)
+    if (isInputRange!R && isSomeChar!(ElementEncodingType!R) &&
```
Any reason `readLink` uses `ElementEncodingType` while `symlink` uses `ElementType`? It doesn't make a difference when passed to `isSomeChar`, but please settle on one.

JakobOvrum commented on std/file.d:
```diff
@@ -2219,15 +2219,31 @@ unittest
+    else
     {
-        return core.sys.posix.unistd.symlink(path1.tempCString(),
-                                             path2.tempCString());
```
It made sense that the calls to `tempCString` were within the `@trusted` block before, despite `tempCString` itself being mistakenly `@trusted`. But I guess the unit tests would ensure that this is properly rectified if we decide to fix `tempCString`.

aG0aep6G replied:
The `tempCString` calls must not be `@trusted`. The types on which `tempCString` is called were more restricted before: only arrays, i.e. no user code. Now it's called on unverified, user-supplied ranges. Calling it in an `@trusted` block would make unverified code `@trusted`.

JakobOvrum replied:
Right, of course.

JakobOvrum commented on std/file.d:
```diff
@@ -2325,6 +2350,15 @@ else version(Posix) string readLink(C)(const(C)[] link) @safe
 }
 
+/// ditto
+version(StdDdoc) string readLink(R)(R link)
```
So `readLink` uses an overload to deal with this, while `symlink` deals with the same problem using just one function. Why the disparity?

aG0aep6G commented:
The differences were just because I had done the two separately and squashed the commits later. Now using `ElementEncodingType` and one overload + `static if` for both. Also added those spaces.

aG0aep6G commented:
Last activity was a week ago. Can I get some more review on this?

DmitryOlshansky commented:
Please rebase

aG0aep6G commented:
> Please rebase

done

DmitryOlshansky commented:
> done

And rebase again ...

aG0aep6G commented:
> And rebase again ...

Done again.

DmitryOlshansky commented:
LGTM

DmitryOlshansky commented:
> Done again.

Ehm.. rebase again

DmitryOlshansky commented on std/file.d:
```diff
@@ -2297,44 +2319,55 @@ version(Posix) @safe unittest
+        auto dynamicBuffer = new char[](bufferLen * 3 / 2);
 
-    auto dynamicBuffer = new char[](bufferLen * 3 / 2);
+        foreach(i; 0 .. 10)
```
Why 10 ? And then what? At least a comment could help a bit

aG0aep6G replied:
Ditto. I just indented this.

DmitryOlshansky commented on std/file.d:
```diff
@@ -2297,44 +2319,55 @@ version(Posix) @safe unittest
+        alias posixReadlink = core.sys.posix.unistd.readlink;
+        enum bufferLen = 2048;
+        enum maxCodeUnits = 6;
+        char[bufferLen] buffer;
```
How about a test to cover at least > 2048 path case?

aG0aep6G replied:
I didn't write this, just indented it. Let's fix pre-existing issues separately.

aG0aep6G commented:
> Ehm.. rebase again

Done again. The conflicts have only been in `changelog.dd` so far.

DmitryOlshansky commented:
Auto-merge toggled on

wilzbach commented:
> The conflicts have only been in changelog.dd so far.

-> hopefully soon fixed by #4228
